dss/api/subscriptions.py

```diff
@@ -3,35 +3,176 @@
 
+
+@dss_handler
 def put(extras: dict, replica: str):
-    pass
+    uuid = str(uuid4())
+    query = extras['query']
+    owner = request.token_info['email']
+
+    es_client = ElasticsearchClient.get(logger)
+
+    index_mapping = {
+        "mappings": {
+            DSS_ELASTICSEARCH_SUBSCRIPTION_TYPE: {
+                "properties": {
+                    "owner": {
+                        "type": "string",
+                        "index": "not_analyzed"
+                    }
+                }
+            }
+        }
+    }
+    # Elasticsearch preprocesses inputs by splitting strings on punctuation.
+    # So for john@example.com, if I searched for people with the email address jill@example.com,
+    # john@example.com would show up because elasticsearch matched example w/ example.
+    # By including "index": "not_analyzed", Elasticsearch leaves all owner inputs alone.
+    get_elasticsearch_index(es_client, DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME, logger, index_mapping)
+
+    try:
+        percolate_registration = _register_percolate(es_client, uuid, query)
+        logger.debug(f"Percolate query registration succeeded:\n{percolate_registration}")
+    except ElasticsearchException:
+        logger.critical(f"Percolate query registration failed:\n{percolate_registration}")
+        raise DSSException(requests.codes.internal_server_error,
+                           "elasticsearch_error",
+                           "Unable to register elasticsearch percolate query!")
+
+    extras['owner'] = owner
+
+    try:
+        subscription_registration = _register_subscription(es_client, uuid, extras)
+        logger.debug(f"Event Subscription succeeded:\n{subscription_registration}")
+    except ElasticsearchException:
+        logger.critical(f"Event Subscription failed:\n{subscription_registration}")
+
+        # Delete percolate query to make sure queries and subscriptions are in sync.
+        es_client.delete(index=DSS_ELASTICSEARCH_INDEX_NAME,
+                         doc_type=DSS_ELASTICSEARCH_QUERY_TYPE,
+                         id=uuid,
+                         refresh=True)
+        raise DSSException(requests.codes.internal_server_error,
+                           "elasticsearch_error",
+                           "Unable to register subscription! Rolling back percolate query.")
 
+    return jsonify(dict(uuid=uuid)), requests.codes.created
 
+
+@dss_handler
 def delete(uuid: str, replica: str):
-    pass
+    owner = request.token_info['email']
+
+    es_client = ElasticsearchClient.get(logger)
+
+    try:
+        response = es_client.get(index=DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME,
+                                 doc_type=DSS_ELASTICSEARCH_SUBSCRIPTION_TYPE,
+                                 id=uuid)
+    except NotFoundError as ex:
+        raise DSSException(requests.codes.not_found, "not_found", "Cannot find subscription!")
+
+    source = response['_source']
+
+    if source['owner'] != owner:
```

Comment:
👍  Changed `owner` to `authenticated_user_email` and `source` to `stored_metadata`